### ISSUE TITLE
[TASK] Check if subscription manager behaves atomically or not

### ISSUE THREAD

**Issue by MericFeyz** [IndustryFusion/DigitalTwin#653]

**Comment by wagmarcel** [comment]
Test: 
1. Stop the NGSI-LD-Update bridge to avoid duplications
1. Create two objects and subscribe to their changes
2. Update the objects in two update rest calls and watch the subscription behavior (e.g. is it updated iteratively or in one step?)j
3. Update the object in one batch REST call and watch the subscription behavior(there should be only one single update)
4. Do the same test with two properties in the same object

**Comment by MericFeyz** [comment]
Unfortunately Scorpio v5.0.5 does not behave atomically.
### Experiment 1
- Create 2 Entities in two individual create requests
- Subscribe to both of them with one subscription observing all entities of their type
- Update 2 Entities in two individual update requests
- Save the notifications to a temp file
This experiment results in 2 consequent notifications:
```
{"body":{"id":"notification:-8443402073510220638","type":"Notification","subscriptionId":"urn:subscription-test:1","notifiedAt":"2025-01-10T13:51:28.764000Z","data":[{"id":"urn:plasmacutter-test:12345","type":"https://industry-fusion.com/types/v0.9/plasmacutter_test","https://industry-fusion.com/types/v0.9/state":{"type":"Property","datasetId":"urn:cutter:test1","value":"OFF"}}]}}
{"body":{"id":"notification:-8264387462205091685","type":"Notification","subscriptionId":"urn:subscription-test:1","notifiedAt":"2025-01-10T13:51:30.825000Z","data":[{"id":"urn:plasmacutter-test:123456","type":"https://industry-fusion.com/types/v0.9/plasmacutter_test","https://industry-fusion.com/types/v0.9/state":{"type":"Property","datasetId":"urn:cutter:test12","value":"OFFF"}}]}}
```
### Experiment 2
- Create 2 Entities in two individual create requests
- Subscribe to both of them with one subscription observing all entities of their type
- Update 2 Entities in one batch update request
- Save the notifications to a temp file
This experiment results in 2 consequent notifications(note that even though the notifiedAt dateTimes are very close to each other, are not identical) :
```
{"body":{"id":"notification:-5635763728599355493","type":"Notification","subscriptionId":"urn:subscription-test:1","notifiedAt":"2025-01-10T13:55:01.909000Z","data":[{"id":"urn:plasmacutter-test:123456","type":"https://industry-fusion.com/types/v0.9/plasmacutter_test","https://industry-fusion.com/types/v0.9/state":{"type":"Property","datasetId":"urn:cutter:test12","value":"OFFF"}}]}}
{"body":{"id":"notification:-5406380717777676753","type":"Notification","subscriptionId":"urn:subscription-test:1","notifiedAt":"2025-01-10T13:55:01.916000Z","data":[{"id":"urn:plasmacutter-test:12345","type":"https://industry-fusion.com/types/v0.9/plasmacutter_test","https://industry-fusion.com/types/v0.9/state":{"type":"Property","datasetId":"urn:cutter:test1","value":"OFF"}}]}}
```
### Experiment 3
- Create one Entity in one individual create request
- Subscribe to the entity with one subscription observing all entities of their type
- Update the same Entity's same attribute two times in one batch update request (e.g. e.State = b && e.State = c)
- Save the notifications to a temp file
This experiment results in 2 consequent notifications(note that even though the notifiedAt dateTimes are very close to each other, are not identical) :
```
{"body":{"id":"notification:-7864066569800536915","type":"Notification","subscriptionId":"urn:subscription-test:1","notifiedAt":"2025-01-10T14:11:30.623000Z","data":[{"id":"urn:plasmacutter-test:12345","type":"https://industry-fusion.com/types/v0.9/plasmacutter_test","https://industry-fusion.com/types/v0.9/state":{"type":"Property","datasetId":"urn:cutter:test1","value":"OFF"}}]}}
{"body":{"id":"notification:-8223072282231665073","type":"Notification","subscriptionId":"urn:subscription-test:1","notifiedAt":"2025-01-10T14:11:30.629000Z","data":[{"id":"urn:plasmacutter-test:12345","type":"https://industry-fusion.com/types/v0.9/plasmacutter_test","https://industry-fusion.com/types/v0.9/state":{"type":"Property","datasetId":"urn:cutter:test1","value":"OFFF"}}]}}
```
### Experiment 4
- Create one Entity with two attributes in one individual create request
- Subscribe to the entity with one subscription observing all entities of their type
- Update the Entity's each attribute once in one batch update request (e.g. e.State = b && e.HasFilter = c)
- Save the notifications to a temp file
This experiment results in 2 consequent notifications(note that even though the notifiedAt dateTimes are very close to each other, are not identical):
```
{"body":{"id":"notification:-7764834395620727608","type":"Notification","subscriptionId":"urn:subscription-test:1","notifiedAt":"2025-01-10T14:15:56.594000Z","data":[{"id":"urn:plasmacutter-test:12345","type":"https://industry-fusion.com/types/v0.9/plasmacutter_test","https://industry-fusion.com/types/v0.9/hasFilter":{"type":"Relationship","object":"urn:filter-test:12345"},"https://industry-fusion.com/types/v0.9/state":{"type":"Property","datasetId":"urn:cutter:test1","value":"OFF"}}]}}
{"body":{"id":"notification:-7634862594722572275","type":"Notification","subscriptionId":"urn:subscription-test:1","notifiedAt":"2025-01-10T14:15:56.600000Z","data":[{"id":"urn:plasmacutter-test:12345","type":"https://industry-fusion.com/types/v0.9/plasmacutter_test","https://industry-fusion.com/types/v0.9/hasFilter":{"type":"Relationship","object":"urn:filter-test:1234567"},"https://industry-fusion.com/types/v0.9/state":{"type":"Property","datasetId":"urn:cutter:test1","value":"OFF"}}]}}
```
To sum up, these experiments, especially the experiment 4, are able to show that the Scorpio applies batch merge operations sequentially and **not** atomically. The proof of it is the fact that the first notification of experiment 4 reflects the changes of the ```e.State = b``` change but not the ```e.HasFilter = c```, and that the second notification reflects both of these changes. 

One slight ray of hope is that the Scorpio's minor version update of yesterday mentions fixes regarding notifications. I will conduct these same experiments on v5.0.6 and report.

**Comment by MericFeyz** [comment]
Example notification output after running batch merge operation on 15 entities.

```
{
  "metadata":[
    {
      "blub":"blabla"
    }
  ],
  "body":{
    "id":"notification:-6717099093951874131",
    "type":"Notification",
    "subscriptionId":"urn:subscription:10",
    "notifiedAt":"2025-04-11T12:45:58.819000Z",
    "data":[
      {
        "id":"house99:smartrooms:room5",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":555
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room4",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":444
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room7",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":777
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room6",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":666
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room9",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":999
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room8",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":888
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room10",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":1010
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room1",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":111
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room13",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":1313
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room14",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":1414
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room3",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":333
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room11",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":1111
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room2",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":222
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room12",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":1212
        },
        "updated":{
          "temperature":null
        }
      },
      {
        "id":"house99:smartrooms:room15",
        "type":"Room",
        "deleted":{
          
        },
        "temperature":{
          "type":"Property",
          "value":1515
        },
        "updated":{
          "temperature":null
        }
      }
    ]
  }
}
```

**Comment by MericFeyz** [comment]
{"a":")$%^&","d":34,"e":1744894615970,"i":false,"j":false,"k":false,"m":false,"b":["house99:smartrooms:room1",{"myuniqueuri:temperature":[{"@type":["https://uri.etsi.org/ngsi-ld/Property"],"https://uri.etsi.org/ngsi-ld/hasValue":[{"@value":11111}],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@value":"2025-04-17T12:56:55.970000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@value":"2025-04-17T12:56:55.970000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}]}],"@id":"house99:smartrooms:room1","https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@value":"2025-04-17T12:56:55.970000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}]},{"@id":"house99:smartrooms:room1","@type":["urn:mytypes:room"],"myuniqueuri:temperature":[{"@type":["https://uri.etsi.org/ngsi-ld/Property"],"https://uri.etsi.org/ngsi-ld/hasValue":[{"@value":111}],"https://uri.etsi.org/ngsi-ld/unitCode":[{"@value":"CEL"}],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@type":"https://uri.etsi.org/ngsi-ld/DateTime","@value":"2025-04-17T12:55:21.078000Z"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@type":"https://uri.etsi.org/ngsi-ld/DateTime","@value":"2025-04-17T12:55:29.887000Z"}],"https://uri.etsi.org/ngsi-ld/default-context/providedBy":[{"@type":["https://uri.etsi.org/ngsi-ld/Relationship"],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@type":"https://uri.etsi.org/ngsi-ld/DateTime","@value":"2025-04-17T12:55:21.078000Z"}],"https://uri.etsi.org/ngsi-ld/hasObject":[{"@id":"smartbuilding:house99:sensor19970309"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@type":"https://uri.etsi.org/ngsi-ld/DateTime","@value":"2025-04-17T12:55:21.078000Z"}]}]}],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@type":"https://uri.etsi.org/ngsi-ld/DateTime","@value":"2025-04-17T12:55:21.078000Z"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@type":"https://uri.etsi.org/ngsi-ld/DateTime","@value":"2025-04-17T12:55:29.887000Z"}]},"house99:smartrooms:room2",{"myuniqueuri:temperature":[{"@type":["https://uri.etsi.org/ngsi-ld/Property"],"https://uri.etsi.org/ngsi-ld/hasValue":[{"@value":22222}],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@value":"2025-04-17T12:56:55.970000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@value":"2025-04-17T12:56:55.970000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}]}],"@id":"house99:smartrooms:room2","https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@value":"2025-04-17T12:56:55.970000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}]},{"@id":"house99:smartrooms:room2","@type":["urn:mytypes:room"],"myuniqueuri:temperature":[{"@type":["https://uri.etsi.org/ngsi-ld/Property"],"https://uri.etsi.org/ngsi-ld/hasValue":[{"@value":222}],"https://uri.etsi.org/ngsi-ld/unitCode":[{"@value":"CEL"}],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@type":"https://uri.etsi.org/ngsi-ld/DateTime","@value":"2025-04-17T12:55:21.080000Z"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@type":"https://uri.etsi.org/ngsi-ld/DateTime","@value":"2025-04-17T12:55:29.887000Z"}],"https://uri.etsi.org/ngsi-ld/default-context/providedBy":[{"@type":["https://uri.etsi.org/ngsi-ld/Relationship"],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@type":"https://uri.etsi.org/ngsi-ld/DateTime","@value":"2025-04-17T12:55:21.080000Z"}],"https://uri.etsi.org/ngsi-ld/hasObject":[{"@id":"smartbuilding:house99:sensor19960913"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@type":"https://uri.etsi.org/ngsi-ld/DateTime","@value":"2025-04-17T12:55:21.080000Z"}]}]}],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@type":"https://uri.etsi.org/ngsi-ld/DateTime","@value":"2025-04-17T12:55:21.080000Z"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@type":"https://uri.etsi.org/ngsi-ld/DateTime","@value":"2025-04-17T12:55:29.887000Z"}]}]}

**Comment by MericFeyz** [comment]
null    {"a":")$%^&","d":31,"e":1744896040524,"i":false,"j":false,"k":false,"m":false,"b":["house99:smartrooms:room1",{"myuniqueuri:temperature":[{"https://uri.etsi.org/ngsi-ld/default-context/providedBy":[{"https://uri.etsi.org/ngsi-ld/hasObject":[{"@id":"smartbuilding:house99:sensor19970309"}],"@type":["https://uri.etsi.org/ngsi-ld/Relationship"],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@value":"2025-04-17T13:20:40.524000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@value":"2025-04-17T13:20:40.524000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}]}],"@type":["https://uri.etsi.org/ngsi-ld/Property"],"https://uri.etsi.org/ngsi-ld/unitCode":[{"@value":"CEL"}],"https://uri.etsi.org/ngsi-ld/hasValue":[{"@value":22}],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@value":"2025-04-17T13:20:40.524000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@value":"2025-04-17T13:20:40.524000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}]}],"@id":"house99:smartrooms:room1","@type":["urn:mytypes:room"],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@value":"2025-04-17T13:20:40.524000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@value":"2025-04-17T13:20:40.524000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}]},null,"house99:smartrooms:room2",{"myuniqueuri:temperature":[{"https://uri.etsi.org/ngsi-ld/default-context/providedBy":[{"https://uri.etsi.org/ngsi-ld/hasObject":[{"@id":"smartbuilding:house99:sensor19960913"}],"@type":["https://uri.etsi.org/ngsi-ld/Relationship"],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@value":"2025-04-17T13:20:40.524000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@value":"2025-04-17T13:20:40.524000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}]}],"@type":["https://uri.etsi.org/ngsi-ld/Property"],"https://uri.etsi.org/ngsi-ld/unitCode":[{"@value":"CEL"}],"https://uri.etsi.org/ngsi-ld/hasValue":[{"@value":23}],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@value":"2025-04-17T13:20:40.524000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@value":"2025-04-17T13:20:40.524000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}]}],"@id":"house99:smartrooms:room2","@type":["urn:mytypes:room"],"https://uri.etsi.org/ngsi-ld/createdAt":[{"@value":"2025-04-17T13:20:40.524000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}],"https://uri.etsi.org/ngsi-ld/modifiedAt":[{"@value":"2025-04-17T13:20:40.524000Z","@type":"https://uri.etsi.org/ngsi-ld/DateTime"}]},null]}
